Add support for 128b atomics to `atomic_ref`

wmaxey opened the pull request:
## Description

When possible, this PR adds support for native 128b atomics to `atomic_ref`.

### Where and what is available?

#### Host `cuda::atomic_ref<int128>`
```
Host compilers:
MSVC - ❌
Clang/GCC - 🟨 (libatomic or intrinsic via override `CCCL_ENABLE_EXPERIMENTAL_HOST_ATOMICS_128B`)
```

#### Device `cuda::atomic_ref<int128>`
```
GPU Architectures:
SM70+ - 128b ld/st supported
SM90+ - all functions enabled. fetch_op functions are emulated via CAS loop.
```


## Checklist
<!-- TODO: - [ ] I am familiar with the [Contributing Guidelines](). -->
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

wmaxey committed:
Add a test for 128b atomic_ref

wmaxey committed:
Fix issues in 128b paths of codegen

wmaxey committed:
Plumb support for 128b into the interface of atomic_ref

wmaxey commented on libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_derived.h:
```diff
@@ -339,6 +339,98 @@ _CCCL_DEVICE _Tp __atomic_exchange_n_cuda(_Tp volatile* __ptr, _Tp __val, int __
+      return __old | __val;
+    },
+    __memorder,
+    _Sco{});
+}
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_bitwise<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_or_cuda(volatile _Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(
+    __ptr,
+    [__val](_Tp __old) {
+      return __old | __val;
+    },
+    __memorder,
+    _Sco{});
+}
+
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_bitwise<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_xor_cuda(_Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(
+    __ptr,
+    [__val](_Tp __old) {
+      return __old ^ __val;
+    },
+    __memorder,
+    _Sco{});
+}
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_bitwise<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_xor_cuda(volatile _Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(
+    __ptr,
+    [__val](_Tp __old) {
+      return __old ^ __val;
+    },
+    __memorder,
+    _Sco{});
+}
+
```
TL;DR: There is no PTX for adding/manipulating 128b values atomically so they're all CAS loop emulated.

wmaxey committed:
Fix mistake in arch size query for atomic_ref, this obviously doesn't work in NVC++.

miscco commented on libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_derived.h:
```diff
@@ -339,6 +339,98 @@ _CCCL_DEVICE _Tp __atomic_exchange_n_cuda(_Tp volatile* __ptr, _Tp __val, int __
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_arithmetic<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_add_cuda(_Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(
+    __ptr,
+    [__val](_Tp __old) {
+      return __old + __val;
+    },
```
This will instantiate a struct everytime this function is instantiated, which is going to be quite often.

We need to pull those lambdas out into a struct that has the right operators

wmaxey replied:
Done, though you might want to see if my plan has any issues. I don't think it will be much improved.

miscco commented on libcudacxx/include/cuda/std/__atomic/types.h:
```diff
@@ -13,6 +13,8 @@
 
 #include <cuda/std/detail/__config>
 
+#include "nv/detail/__target_macros"
+
```
Please add it below at the right place
```suggestion
```

wmaxey replied:
... Some extension thought I needed that apparently. 😮‍💨

wmaxey committed:
Merge branch 'main' into fea/atomic_ref_128

wmaxey committed:
Refactor atomic ops to use common structs

wmaxey committed:
Fixup extra include

wmaxey committed:
Cleanup comments

wmaxey committed:
precommit fixup

wmaxey committed:
Remove hacky test for compatibility

wmaxey committed:
Disable test on Windows

wmaxey committed:
Allow passing in --device-int128 to NVRTCC

wmaxey committed:
Add warnings and macro controls for int128 support.

Users can use `#define LIBCUDACXX_IGNORE_MISSING_BUILTIN_128_ATOMICS` to
disable checking on GCC and allowing dispatch through libatomic.

wmaxey committed:
Flush stdout stream in libcu++ tests before running device code

wmaxey committed:
White space fixup

wmaxey committed:
Add tests verifying behavior of atomic 128b control macros

wmaxey committed:
Use the callable macro convention for testing 128b support

wmaxey committed:
Remove CMake controls for setting 'mcx16'

wmaxey committed:
Add heterogeneous tests for large atomics

wmaxey committed:
Fixup warnings for protected data members in small atomic codepath

wmaxey committed:
Make atomic_fetch.fail.cpp specifically about half and prepare it for future changes

wmaxey committed:
Link libatomic from CMake

wmaxey committed:
Fixup in atomic test for missing library warning override

wmaxey committed:
Merge remote-tracking branch 'origin/main' into fea/atomic_ref_128

wmaxey committed:
Use extern 'C' to generate errors instead of static_assert for atomic_ref<int128>

wmaxey committed:
block 16b atomic_ref on platforms with no support

wmaxey committed:
Merge remote-tracking branch 'origin/main' into fea/atomic_ref_128

wmaxey committed:
Touch ups on isa detection

wmaxey committed:
Fixup fallback for atomic128 host support in _config.

wmaxey committed:
Restrict small class warning pragma to GCC8 and up

wmaxey committed:
Merge branch 'main' into fea/atomic_ref_128

wmaxey committed:
Merge branch 'main' into fea/atomic_ref_128

wmaxey committed:
Spelling fixup

wmaxey committed:
Disallow pre-sm70 paths of atomic_ref 128b ld/st

wmaxey committed:
Add a 16b atomic ld/st test for sm70+

wmaxey committed:
Document atomic_ref with 16b types

wmaxey committed:
Revert branch nvrtc changes

wmaxey committed:
Merge branch 'main' into fea/atomic_ref_128

wmaxey committed:
Fix swapped mov instructions in 128b codegen

wmaxey committed:
Merge remote-tracking branch 'origin/main' into fea/atomic_ref_128

wmaxey committed:
Change 128b tests to skip host testing

wmaxey committed:
Change atomics to only work in host if specifically enabled

wmaxey committed:
Add check for 128b constructible atomic

wmaxey committed:
Test fixups

wmaxey committed:
Rebuild generated ptx file

miscco commented on docs/libcudacxx/extended_api/synchronization_primitives/atomic_ref.rst:
```diff
@@ -38,13 +38,17 @@ with the following additional operations. This class additionally deviates from
 
 For ``cuda::atomic_ref<T>`` and ``cuda::std::atomic_ref<T>`` the type ``T`` must satisfy the following:
-  - ``sizeof(T) <= 8``.
+  - ``sizeof(T) <= 8 // (16 for SM70+)``.
```
I find it confusing that we specify this differently than above. We should use the same description

wmaxey replied:
I think this is resolved.

miscco commented on libcudacxx/include/cuda/std/__atomic/types/small.h:
```diff
@@ -63,9 +63,15 @@ _CCCL_HOST_DEVICE inline __atomic_small_proxy_t<_Tp> __atomic_small_to_32(_Tp __
 template <class _Tp, enable_if_t<!is_arithmetic_v<_Tp>, int> = 0>
 _CCCL_HOST_DEVICE inline _Tp __atomic_small_from_32(__atomic_small_proxy_t<_Tp> __val)
 {
+  _CCCL_DIAG_PUSH
+#if _CCCL_COMPILER(GCC, >=, 8)
+  // GCC starting with GCC8 warns about our extended floating point types having protected data members
+  _CCCL_DIAG_SUPPRESS_GCC("-Wclass-memaccess")
+#endif
```
Paranoia: NVHPC has issues with warning suppression within classes 

Can we pull that diagnostic out of the function?

wmaxey replied:
Moved outside of function scope.

miscco commented on libcudacxx/include/cuda/std/detail/libcxx/include/__config:
```diff
@@ -146,11 +146,24 @@ extern "C++" {
 #  endif
 
-#  define _LIBCUDACXX_ATOMIC_UNSAFE_AUTOMATIC_STORAGE
+// Clang provides 128b atomics as a builtin
+#  if defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_HOST_ATOMICS_128B)
+#    define _CCCL_HOST_128_ATOMICS_ENABLED() 1
+#    define _CCCL_HOST_128_ATOMICS_MAYBE()   0
+// GCC does not provide 128b atomics, but they may be available as a library, this requires opt-in usage.
+// See: https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html "-mcx16" for more
+#  elif _CCCL_COMPILER(CLANG) || _CCCL_COMPILER(GCC)
+#    define _CCCL_HOST_128_ATOMICS_ENABLED() 0
+#    define _CCCL_HOST_128_ATOMICS_MAYBE()   1
+#  else
+#    define _CCCL_HOST_128_ATOMICS_ENABLED() 0
+#    define _CCCL_HOST_128_ATOMICS_MAYBE()   0
+#  endif
+
```
We have a `__internal/features.h` I believe that would be the appropriate place for this

wmaxey replied:
Moved.

wmaxey committed:
Merge remote-tracking branch 'origin/main' into fea/atomic_ref_128

wmaxey commented on libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_derived.h:
```diff
@@ -341,50 +366,74 @@ _CCCL_DEVICE _Tp __atomic_exchange_n_cuda(_Tp volatile* __ptr, _Tp __val, int __
+template <typename _Tp, typename _Up, typename _Sco, __atomic_enable_if_not_native_arithmetic<_Tp> = 0>
+_CCCL_DEVICE _Tp __atomic_fetch_add_cuda(_Tp* __ptr, _Up __val, int __memorder, _Sco)
+{
+  return __atomic_fetch_update_cuda(__ptr, __cuda_atomic_op_bind<_Tp, _CUDA_VSTD::plus>{__val}, __memorder, _Sco{});
```
These aren't likely to be ambiguous, and I may address this another time.

wmaxey commented:
> would not be better to use [NVVM intrinsics](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#nv-atomic-exchange) where possible to mitigate compilation time?

Later, yes, we need to develop a header for using those when the appropriate CTK version is detected.

wmaxey committed:
Drop _CUDA_VSTD.

wmaxey committed:
Address review comments in documentation.

wmaxey committed:
Change `LIBCUDACXX` flag to `CCCL`.

wmaxey committed:
Address memaccess warning being inside function.

wmaxey committed:
Merge branch 'main' into fea/atomic_ref_128

wmaxey committed:
`[[nodiscard]]` annotations.

wmaxey committed:
fixup nodiscard location.

wmaxey committed:
More attributes on atomic functions

wmaxey committed:
CCCL_DEVICE -> CCCL_DEVICE_API

wmaxey committed:
Revert "CCCL_DEVICE -> CCCL_DEVICE_API"

This reverts commit 8f1b5b33dfa5816bffb1af7719139ea2b1aa7d8d.

miscco committed:
Fix formatting

miscco committed:
Merge branch 'main' into fea/atomic_ref_128

gonidelis reviewed (commented):
Few nits as long as we make sure alignment requirements are clear and do not lead to undefined behavior on type casts should be ok. - _apologies for the multi review panels I used the VS Code GH PRs extension for the first time_

gonidelis commented on libcudacxx/codegen/generators/ld_st.h:
```diff
@@ -79,11 +79,18 @@ static inline _CCCL_DEVICE void __cuda_atomic_load_memory_order_dispatch(_Fn &__
 static inline _CCCL_DEVICE void __cuda_atomic_load(
   const _Type* __ptr, _Type& __dst, {3}, __atomic_cuda_operand_{0}{1}, {5}, {7})
 {{
+  static_assert(__cccl_ptx_isa >= 840 && (sizeof(_Type) == 16), "128b ld/st is not supported until PTX ISA version 840");
```
`__cccl_ptx_isa >= 840 ` 
SM70 too for ld/st, no?

wmaxey replied:
We only support SM70+ so we don't need to guard it for earlier architectures.

gonidelis reviewed (commented):
https://github.com/NVIDIA/cccl/blob/c00a16e27a54fb5342e6bcc3d51ec42197f809db/libcudacxx/codegen/generators/fetch_ops.h#L148

Add `128` here?

gonidelis reviewed (changes requested):
see above

wmaxey commented on docs/libcudacxx/extended_api/synchronization_primitives/atomic_ref.rst:
```diff
@@ -38,13 +38,17 @@ with the following additional operations. This class additionally deviates from
+  - ``sizeof(T) <= 16``.
   - ``T`` must not have “padding bits”, i.e., T's `object representation <https://en.cppreference.com/w/cpp/language/object#Object_representation_and_value_representation>`_
     must not have bits that do not participate in it's value representation.
 
```
That's probably a good idea.

wmaxey commented on libcudacxx/codegen/generators/exchange.h:
```diff
@@ -61,13 +61,20 @@ template <class _Type>
 static inline _CCCL_DEVICE void __cuda_atomic_exchange(
   _Type* __ptr, _Type& __old, _Type __new, {4}, __atomic_cuda_operand_{0}{1}, {6})
 {{
+  static_assert(__cccl_ptx_isa >= 840 && (sizeof(_Type) == 16), "128b CAS is not supported until PTX ISA version 840");
```
Changed.

wmaxey commented on libcudacxx/codegen/generators/exchange.h:
```diff
@@ -61,13 +61,20 @@ template <class _Type>
+  static_assert(__cccl_ptx_isa >= 840 && (sizeof(_Type) == 16), "128b CAS is not supported until PTX ISA version 840");
+  NV_DISPATCH_TARGET(
+    NV_PROVIDES_SM_90, (),
+    NV_ANY_TARGET, (__atomic_cas_128b_unsupported_before_SM_90();)
```
Ditto

wmaxey committed:
Clarification CAS->exchange for some functions

wmaxey committed:
Add note about alignment

wmaxey committed:
Merge remote-tracking branch 'origin/main' into fea/atomic_ref_128

wmaxey committed:
Merge branch 'main' into fea/atomic_ref_128

fbusato reviewed (approved):
looks good. We need to consider atomic intrinsics in the future to mitigate compile-time and compose the API with `annotated_ptr`
